lib/Dialect/Lattigo/Transforms/ConfigureCryptoContext.cpp

```diff
@@ -0,0 +1,139 @@
+    return WalkResult::advance();
+  });
+  SmallVector<int64_t> rotIndicesResult(distinctRotIndices.begin(),
+                                        distinctRotIndices.end());
+  return rotIndicesResult;
+}
+
+LogicalResult convertFunc(func::FuncOp op) {
+  auto module = op->getParentOfType<ModuleOp>();
+  std::string configFuncName("");
+  llvm::raw_string_ostream configNameOs(configFuncName);
+  configNameOs << op.getSymName() << "__configure";
+
+  ImplicitLocOpBuilder builder =
+      ImplicitLocOpBuilder::atBlockEnd(module.getLoc(), module.getBody());
+
+  // __configure() -> (evaluator, params, encoder, encryptor, decryptor)
+  SmallVector<Type> funcArgTypes;
+  SmallVector<Type> funcResultTypes;
+
+  Type evaluatorType = BGVEvaluatorType::get(builder.getContext());
+  Type paramsType = BGVParameterType::get(builder.getContext());
+  Type encoderType = BGVEncoderType::get(builder.getContext());
+  Type encryptorType = RLWEEncryptorType::get(builder.getContext());
+  Type decryptorType = RLWEDecryptorType::get(builder.getContext());
+  funcResultTypes.push_back(evaluatorType);
+  funcResultTypes.push_back(paramsType);
+  funcResultTypes.push_back(encoderType);
+  funcResultTypes.push_back(encryptorType);
+  funcResultTypes.push_back(decryptorType);
+
+  FunctionType configFuncType =
+      FunctionType::get(builder.getContext(), funcArgTypes, funcResultTypes);
+
+  auto configFuncOp =
+      builder.create<func::FuncOp>(configFuncName, configFuncType);
+  builder.setInsertionPointToEnd(configFuncOp.addEntryBlock());
+
+  // TODO: Allow custom params
+  // 128-bit secure parameters enabling depth-7 circuits.
+  // LogN:14, LogQP: 431.
+  auto logN = 14;
+  auto paramAttr = BGVParametersLiteralAttr::get(
+      builder.getContext(), /*logN*/ logN, /*Q*/ nullptr, /*P*/ nullptr,
+      /*logQ*/
+      DenseI32ArrayAttr::get(builder.getContext(),
+                             {55, 45, 45, 45, 45, 45, 45, 45}),
+      /*logP*/ DenseI32ArrayAttr::get(builder.getContext(), {61}),
+      /*ptm*/ 0x10001);
+  auto paramType = BGVParameterType::get(builder.getContext());
+
+  auto params =
+      builder.create<BGVNewParametersFromLiteralOp>(paramType, paramAttr);
+
+  auto encoder = builder.create<BGVNewEncoderOp>(encoderType, params);
+  auto kgenType = RLWEKeyGeneratorType::get(builder.getContext());
+  auto kgen = builder.create<RLWENewKeyGeneratorOp>(kgenType, params);
+  auto skType = RLWESecretKeyType::get(builder.getContext());
+  auto pkType = RLWEPublicKeyType::get(builder.getContext());
+  SmallVector<Type> keypairTypes = {skType, pkType};
+  auto keypair = builder.create<RLWEGenKeyPairOp>(keypairTypes, kgen);
+  auto sk = keypair.getResult(0);
+  auto pk = keypair.getResult(1);
+  auto encryptor =
+      builder.create<RLWENewEncryptorOp>(encryptorType, params, pk);
+  auto decryptor =
+      builder.create<RLWENewDecryptorOp>(decryptorType, params, sk);
+
+  // TODO: do not always generate Relinearization Key
```

Comment:
Would you just be able to check for multiplications? Or is the issue that the ops (evalkeyset, etc) require the relinearization key?

Reply:
> Or is the issue that the ops (evalkeyset, etc) require the relinearization key?

Yes currently NewEvaluationKeySetOp requires a relinearization key.

```tablegen
    Lattigo_RLWERelinearizationKey:$relinearizationKey,
    Variadic<Lattigo_RLWEGaloisKey>:$galoisKeys
```
I could not make `$relinearizationKey` optional, as MLIR would then complain

```
error: op has multiple variadic operands but no specification over their sizes
```

Reply:
I see, thank you! Fine with me for now, I think a trait like AttrSizedOperandSegments /may/ help (attaches an operand size attribute for each size) later